src/utils/gauges.ts

```diff
@@ -372,3 +430,171 @@ export function areGaugesActive(chainId?: ChainId): boolean {
     IS_VESDL_LIVE
   )
 }
+
+async function getGaugeCount(
+  childGaugeFactory: ChildGaugeFactory | null,
+  gaugeController: GaugeController | null,
+): Promise<BigNumber> {
+  if (childGaugeFactory) {
+    return await childGaugeFactory.get_gauge_count()
+  }
+
+  if (gaugeController) {
+    return await gaugeController.n_gauges()
+  }
+
+  return Promise.resolve(Zero)
+}
+
+function retrieveGaugeAddresses(
+  gaugeControllerMultiCall: MulticallContract<GaugeController>,
+  childGaugeFactoryMultiCall: MulticallContract<ChildGaugeFactory>,
+): (
+  value: number,
+  index: number,
+  array: number[],
+) => MulticallCall<
+  [arg0: BigNumberish, overrides?: CallOverrides | undefined],
+  string
+> {
+  if (gaugeControllerMultiCall.address) {
+    return (value) => gaugeControllerMultiCall.gauges(value)
+  }
+  return (value) => childGaugeFactoryMultiCall.get_gauge(value)
+}
+
+const retrieveGaugeContracts = (
+  gaugeAddresses: string[],
+  chainId: ChainId,
+): MulticallContract<LiquidityGaugeV5>[] | MulticallContract<ChildGauge>[] => {
+  if (chainId === ChainId.MAINNET) {
+    return gaugeAddresses.map((address) =>
+      createMultiCallContract<LiquidityGaugeV5>(
+        address,
+        LIQUIDITY_GAUGE_V5_ABI,
+      ),
+    )
+  }
+
+  return gaugeAddresses.map((address) =>
+    createMultiCallContract<ChildGauge>(address, CHILD_GAUGE_ABI),
+  )
+}
+
+async function getSDLRates(
+  gaugeMinterContract: Minter | null,
+  gaugeRelativeWeights: BigNumber[],
+  gaugeMulticallContracts: MulticallContract<ChildGauge>[],
+  ethCallProvider: MulticallProvider,
+) {
+  if (gaugeMinterContract) {
+    const minterSDLRate = await gaugeMinterContract.rate()
+    return gaugeRelativeWeights.map((weight) =>
+      minterSDLRate.mul(weight).div(BN_1E18),
+    )
+  }
+
+  const currentTimeStamp = Date.now() / 1000
+  return await ethCallProvider.tryAll(
+    gaugeMulticallContracts.map((contract) =>
+      contract.inflation_rate(Math.floor(currentTimeStamp / 604800)),
+    ),
+  )
+}
+
+async function getGaugeRewardCounts(
+  gaugeMulticallContracts:
+    | MulticallContract<LiquidityGaugeV5>[]
+    | MulticallContract<ChildGauge>[],
+  ethCallProvider: MulticallProvider,
+) {
+  return (
+    await ethCallProvider.tryAll(
+      gaugeMulticallContracts.map((contract) => contract.reward_count()),
+    )
+  ).map((count) => count || Zero)
+}
+
+async function getGaugeRewardTokens(
+  gaugeMulticallContracts: (
+    | MulticallContract<ChildGauge>
+    | MulticallContract<LiquidityGaugeV5>
+  )[],
+  gaugeRewardCounts: BigNumber[],
+  ethCallProvider: MulticallProvider,
+) {
+  return Promise.all(
+    gaugeRewardCounts.map((count, index) =>
+      ethCallProvider.all(
+        enumerate(count.toNumber(), 0).map((num) =>
+          gaugeMulticallContracts[index].reward_tokens(num),
```

Comment:
✅

Reply:
Heads up. Since RootGauge lacks most of these functions relating to rewards, these calls will fail.

Reply:
Hmm, gotta go back to the drawing board to see how I could handle `RootGauges`  and `LiquidityV5Gauges`.  we would only want to calculate rewards for mainnet gauges since we don't allow users to claim sidechain gauges (represented by rootGauges) while on mainnet.

Looking for a way to discern between the 2 types of gauges.

Reply:
https://github.com/saddle-finance/saddle-contract/blob/923e8a398bcdf5c1144e214f883f5f27b9f525cd/contracts/xchainGauges/implementations/RootGauge.vy#L47-L49

One solution is to use `tryAll()` and attempt to read one of these values. 

Only root gauges will return valid values while liquidity v5 will be reverted (returns null on multicall)